misc/pytorch_toolkit/chest_xray_screening/chest_xray_screening/inference.py

```diff
@@ -0,0 +1,184 @@
+        out_pred = torch.FloatTensor().to(self.device)
+        with torch.no_grad():
+            for var_input, var_target in self.data_loader_test:
+                var_target = var_target.to(self.device)
+                var_input = var_input.to(self.device)
+                out_gt = torch.cat((out_gt, var_target), 0).to(self.device)
+
+                _, c, h, w = var_input.size()
+                var_input = var_input.view(-1, c, h, w)
+                out = self.model(var_input.to(self.device))
+                out_pred = torch.cat((out_pred, out), 0)
+        auroc_individual = compute_auroc(tfunc.one_hot(out_gt.squeeze(1).long()).float(), out_pred, self.class_count)
+        auroc_mean = np.array(auroc_individual).mean()
+        print(f'AUROC mean: {auroc_mean}')
+        for i, auroc_val in enumerate(auroc_individual):
+            print(f"{self.class_names[i]}:{auroc_val}")
+        return auroc_mean
+
+    def test_onnx(self, img_path, onnx_checkpoint):
+        onnx_model = onnx.load(onnx_checkpoint)
+        onnx.checker.check_model(onnx_model)
+        # The validity of the ONNX graph is verified by checking the model’s version,
+        # the graph’s structure, as well as the nodes and their inputs and outputs.
+        ort_session = onnxruntime.InferenceSession(onnx_checkpoint)
+        sample_image = Image.open(img_path).convert('RGB')
+        to_tensor = transforms.ToTensor()
+        sample_image = to_tensor(sample_image)
+        sample_image.unsqueeze_(0)
+        ort_inputs = {ort_session.get_inputs()[0].name: to_numpy(sample_image)}
+        ort_outs = ort_session.run(None, ort_inputs)
+        torch_out = self.model(sample_image.cuda())
+        np.testing.assert_allclose(to_numpy(torch_out), ort_outs[0], rtol=1e-03, atol=1e-05)
+
+    def validate_models(self, run_type, onnx_checkpoint =''):
+        cudnn.benchmark = True
+        out_gt = torch.FloatTensor()
+        out_pred = torch.FloatTensor()
+        with torch.no_grad():
+            for var_input, var_target in self.data_loader_test:
+                if run_type in ('pytorch', 'onnx'):
+                    var_target = var_target.to(self.device)
+                    var_input = var_input.to(self.device)
+                    out_gt = out_gt.to(self.device)
+                    out_pred = out_pred.to(self.device)
+                    out_gt = torch.cat((out_gt, var_target), 0)
+                else:
+                    out_gt = torch.cat((out_gt, var_target), 0)
+                _, c, h, w = var_input.size()
+                var_input = var_input.view(-1, c, h, w)
+                to_tensor = transforms.ToTensor()
+                if run_type == 'pytorch':
+                    out = self.model(var_input)
+                elif run_type == 'onnx':
+                    ort_session = onnxruntime.InferenceSession(onnx_checkpoint)
+                    ort_inputs = {ort_session.get_inputs()[0].name: to_numpy(var_input)}
```

Comment:
There is no need to load the model in every iteration on the datasets, it is slow. You can load the model in advance and then run the inference in the loop. Also, it seems like now there are two functions for doing the same: both `test` and `validate_models` can run the pytorch model. 

I guess, the better option in this case is to store the `run_type` as the class variable and create separate method, which encapsulates the logic of the specific model inference.
I.e., you can create two methods (e.g., load_model and inference_model) and, depending on the run_type, they will act differently but produce the same result (loading model and doing one inference step).

Reply:
> both test and validate_models can run the PyTorch model.
Removed `test` and have made necessary changes.

> you can create two methods (e.g., load_model and inference_model)
I have created separate methods.